### PR TITLE
ARROW-8471: [C++][Integration] Represent 64 bit integers as strings

### PR DESCRIPTION
**Pull request by bkietz**
Some 64 bit integer values are not representable as JSON numbers, so store these as strings.

This is issue is a regression; original fix was #5267

### CODE REVIEW

File: cpp/src/arrow/ipc/json_integration_test.cc [hunk 1]
```diff
@@ -255,7 +255,7 @@ static const char* JSON_EXAMPLE = R"example(
     "fields": [
       {
         "name": "foo",
-        "type": {"name": "int", "isSigned": true, "bitWidth": 32},
+        "type": {"name": "int", "isSigned": true, "bitWidth": 64},
```
Comment (fsaintjacques) [inline cpp/src/arrow/ipc/json_integration_test.cc:258]: Add a i32 column so it shows both behavior.

File: dev/archery/archery/integration/tester_go.py [hunk 2]
```diff
@@ -26,7 +26,8 @@ class GoTester(Tester):
     CONSUMER = True
 
     # FIXME(sbinet): revisit for Go modules
-    GOPATH = os.getenv('GOPATH', '~/go')
+    HOME = os.getenv('HOME', '~')
```
Comment (fsaintjacques) [inline dev/archery/archery/integration/tester_go.py:29]: Why the change, just set `GOPATH` if you have something else?

File: cpp/src/arrow/ipc/json_integration_test.cc [hunk 3]
```diff
@@ -272,7 +272,7 @@ static const char* JSON_EXAMPLE = R"example(
         {
           "name": "foo",
           "count": 5,
-          "DATA": [1, 2, 3, 4, 5],
+          "DATA": ["1", "2", "3", "4", "5"],
```
Comment (fsaintjacques) [inline cpp/src/arrow/ipc/json_integration_test.cc:275]: add literal for INT64_MIN and INT64_MAX.

File: dev/archery/archery/integration/datagen.py [hunk 4]
```diff
@@ -179,8 +179,8 @@ def generate_column(self, size, name=None):
         return self.generate_range(size, lower_bound, upper_bound, name=name)
 
     def generate_range(self, size, lower, upper, name=None):
-        values = [int(x) for x in
-                  np.random.randint(lower, upper, size=size)]
+        values = list(map(int if self.bit_width < 64 else str,
+                          np.random.randint(lower, upper, size=size)))
```
Comment (fsaintjacques) [inline dev/archery/archery/integration/datagen.py:183]: Append the extreme values so we test all integrations.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]

File: go/arrow/internal/arrjson/arrjson.go [hunk 5]
```diff
@@ -963,19 +963,23 @@ func i32ToJSON(arr *array.Int32) []interface{} {
...
+		o[i] = vv
 	}
 	return o
 }
 
 func i64ToJSON(arr *array.Int64) []interface{} {
 	o := make([]interface{}, arr.Len())
 	for i := range o {
-		o[i] = arr.Value(i)
+    if arr.IsValid(i) {
```
Comment (pitrou) [inline go/arrow/internal/arrjson/arrjson.go:978]: Indentation problem here?

File: go/arrow/internal/arrjson/arrjson.go [hunk 6]
```diff
@@ -1043,19 +1047,23 @@ func u32ToJSON(arr *array.Uint32) []interface{} {
...
+		o[i] = vv
 	}
 	return o
 }
 
 func u64ToJSON(arr *array.Uint64) []interface{} {
 	o := make([]interface{}, arr.Len())
 	for i := range o {
-		o[i] = arr.Value(i)
+    if arr.IsValid(i) {
```
Comment (pitrou) [inline go/arrow/internal/arrjson/arrjson.go:1062]: Indentation problem here as well?

File: go/arrow/internal/arrjson/arrjson.go [hunk 7]
```diff
@@ -1205,7 +1213,11 @@ func date64FromJSON(vs []interface{}) []arrow.Date64 {
 func date64ToJSON(arr *array.Date64) []interface{} {
 	o := make([]interface{}, arr.Len())
 	for i := range o {
-		o[i] = int64(arr.Value(i))
+    if arr.IsValid(i) {
```
Comment (pitrou) [inline go/arrow/internal/arrjson/arrjson.go:1216]: And here as well.

File: go/arrow/internal/arrjson/arrjson.go [hunk 8]
```diff
@@ -1245,15 +1257,19 @@ func time64FromJSON(vs []interface{}) []arrow.Time64 {
 func time64ToJSON(arr *array.Time64) []interface{} {
 	o := make([]interface{}, arr.Len())
 	for i := range o {
-		o[i] = int64(arr.Value(i))
+    if arr.IsValid(i) {
```
Comment (pitrou) [inline go/arrow/internal/arrjson/arrjson.go:1260]: And here.

File: go/arrow/internal/arrjson/arrjson.go [hunk 9]
```diff
@@ -1265,7 +1281,11 @@ func timestampFromJSON(vs []interface{}) []arrow.Timestamp {
 func timestampToJSON(arr *array.Timestamp) []interface{} {
 	o := make([]interface{}, arr.Len())
 	for i := range o {
-		o[i] = int64(arr.Value(i))
+    if arr.IsValid(i) {
```
Comment (pitrou) [inline go/arrow/internal/arrjson/arrjson.go:1284]: And here.

File: go/arrow/internal/arrjson/arrjson.go [hunk 10]
```diff
@@ -1330,7 +1350,11 @@ func durationFromJSON(vs []interface{}) []arrow.Duration {
 func durationToJSON(arr *array.Duration) []interface{} {
 	o := make([]interface{}, arr.Len())
 	for i := range o {
-		o[i] = arr.Value(i)
+    if arr.IsValid(i) {
```
Comment (pitrou) [inline go/arrow/internal/arrjson/arrjson.go:1353]: And here.

File: dev/archery/archery/integration/tester_go.py (hunk 2 above)
Comment (bkietz) [reply dev/archery/archery/integration/tester_go.py:29]: I don't know why `~` failed resolve to `$HOME` locally but this seemed like a way to save someone the next person from the frustration of reading the script to figure out that `GOPATH` needs set.

File: cpp/src/arrow/ipc/json_integration_test.cc (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/ipc/json_integration_test.cc:275]: Will do

File: dev/archery/archery/integration/datagen.py (hunk 4 above)
Comment (bkietz) [reply dev/archery/archery/integration/datagen.py:183]: alright

File: cpp/src/arrow/type_traits.h [hunk 11]
```diff
@@ -644,6 +644,14 @@ template <typename T, typename R = void>
 using enable_if_physical_unsigned_integer =
     enable_if_t<is_physical_unsigned_integer_type<T>::value, R>;
 
+template <typename T>
+using is_physical_integer_type =
+    std::integral_constant<bool, is_physical_unsigned_integer_type<T>::value ||
+                                     is_physical_signed_integer_type<T>::value>;
```
Comment (kiszk) [inline cpp/src/arrow/type_traits.h:650]: not sure 100%. but, is this indentation correct?

File: go/arrow/internal/arrjson/arrjson.go (hunk 5 above)
Comment (bkietz) [reply go/arrow/internal/arrjson/arrjson.go:978]: vim was expanding tabs to spaces. Thanks

File: cpp/src/arrow/type_traits.h (hunk 11 above)
Comment (bkietz) [reply cpp/src/arrow/type_traits.h:650]: C++ indentation is deferred to `clang-format`. We have a CI job which ensures that running `clang-format` is a no-op. As for explaining why it made this choice... I'm not familiar enough with its implementation to say, sorry

**Review by fsaintjacques — APPROVED** [review]
Good thing we actually validates the extreme value now.

**Comment by pitrou** [comment]
> @bkietz one thing, you didn't change anything in Java, so either it was already behaving like this, or the test is disabled for it?

Java probably accepts both forms, because AFAICT the test isn't disabled for it.

**Comment by pitrou** [comment]
I just looked at the Java code: for Int64 and UInt64, it uses `jackson.core.JsonParser::getValueAsString`, which accepts both number and string tokens.

**Review by pitrou — APPROVED** [review]
+1

**Comment by pitrou** [comment]
@sbinet Could you validate the Go changes here?

**Comment by fsaintjacques** [comment]
I say we move on and accept the go changes.

**Comment by wesm** [comment]
This patch is failing some of the time in Appveyor due to integer narrowing. I'm fixing the problems in #7352
